### PR TITLE
refactor: StackEnsemble with rule-based ensembler

### PR DESCRIPTION
**Pull request by nhanitvn**
- [x] Add RuleBasedModeler
- [x] Refactor StackEnsembleModeler to accept Modeler instead of Model for the ensembler
- [x] Test cases passed

### CODE REVIEW

File: h1st/model/ensemble/stack_ensemble_modeler.py [hunk 1]
```diff
@@ -55,17 +42,31 @@ def build_model(self, data: Dict[str, Any] = None) -> StackEnsemble:
...
             }
         """
-        if not data:
+        if data is None:
             data = self.load_data()
 
-        ensembler = self.train_base_model(data)
+        if data is not None:
+            X_train, y_train = (data['X_train'], data['y_train'])
+            X_train = self.model_class.preprocess(self._sub_models,
```
Comment (TgithubJ) [inline h1st/model/ensemble/stack_ensemble_modeler.py:50]: @nhanitvn Do you intentionally only use sub models' output as an input for ensembler ? Could we concatenate original input X_train and sub models' output to make the input for ensembler ?

Comment (nhanitvn) [reply h1st/model/ensemble/stack_ensemble_modeler.py:50]: I see. Should we make that behavior an option?

File: h1st/model/ensemble/stack_ensemble.py [hunk 2]
```diff
@@ -22,28 +17,32 @@ def __init__(self, ensembler, sub_models: List[PredictiveModel], **kwargs):
...
+                        submodel_predict_input_key: str,
+                        submodel_predict_output_key: str,
+                        X: Any) -> Any:
         """
         Predicts all sub models then merge input raw data with predictions for ensembler's training or prediction
         :param X: raw input data
         """
         if isinstance(X, list):
-            predictions = [StackEnsemble.get_submodels_prediction(sub_models,
+            predictions = [np.hstack((x, StackEnsemble.get_submodels_prediction(sub_models,
```
Comment (nhanitvn) [inline h1st/model/ensemble/stack_ensemble.py:35]: @TgithubJ I added code to combine input data with predictions here

File: h1st/model/ensemble/stack_ensemble_modeler.py (hunk 1 above)
Comment (nhanitvn) [reply h1st/model/ensemble/stack_ensemble_modeler.py:50]: Updated the preprocess() method to perform the concatenation.

**Review by TgithubJ — APPROVED** [review]

File: h1st/model/ensemble/stack_ensemble.py (hunk 2 above)
Comment (TgithubJ) [reply h1st/model/ensemble/stack_ensemble.py:35]: @nhanitvn Awesome-!

**Pull request merged**